src/helpers/files.js

```diff
@@ -29,18 +60,61 @@ exports.moveStaticPages = async ({ netlifyConfig, target, i18n, failBuild }) =>
     dot: true,
   })
 
+  const matchingMiddleware = new Set()
+  const matchedPages = new Set()
+
   // Limit concurrent file moves to number of cpus or 2 if there is only 1
   const limit = pLimit(Math.max(2, cpus().length))
   const promises = pages.map(async (rawPath) => {
     const filePath = slash(rawPath)
     if (isDynamicRoute(filePath)) {
       return
     }
+    // Middleware matches against the unlocalised path
+    const unlocalizedPath = stripLocale(rawPath, i18n?.locales)
+    const middlewarePath = matchMiddleware(middleware, unlocalizedPath)
+    // If a file matches middleware it can't be offloaded to the CDN, and needs to stay at the origin to be served by next/server
+    if (middlewarePath) {
+      matchingMiddleware.add(middlewarePath)
+      matchedPages.add(rawPath)
+      return
+    }
     return limit(moveFile, filePath)
   })
   await Promise.all(promises)
   console.log(`Moved ${files.length} files`)
 
+  if (matchedPages.size !== 0) {
+    console.log(
+      yellowBright(outdent`Skipped moving ${matchedPages.size} ${
+        matchedPages.size === 1 ? 'file because it matches' : 'files because they match'
+      } middleware, so cannot be deployed to the CDN and will be served from the origin instead. This is fine, but we're letting you know because it may not be what you expect.
+      `),
+    )
+
+    console.log(
+      outdent`
+      
+    The following middleware matched statically-rendered pages:
+
+    ${yellowBright([...matchingMiddleware].map((mid) => `- /${mid}/_middleware`).join('\n'))}
+    
+    `,
+    )
+    // There could potentially be thousands of matching pages, so we don't want to spam the console with this
```

Comment:
Is it fine not to show any pages at all when this happens? Is it worth showing something like "The following files and X others matched middlware (...)", or is that overkill?

Reply:
Hmm, possibly, though I'm not sure if that's useful information.